firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/UpdateApkClient.java

```diff
@@ -57,51 +60,51 @@ public void updateApk(
       @NonNull String downloadUrl,
       @NonNull Activity currentActivity) {
 
-    downloadApk(downloadUrl, updateTask)
+    this.cachedUpdateTask = updateTask;
+    downloadApk(downloadUrl)
         .addOnSuccessListener(
             downloadExecutor,
             file ->
                 install(file.getPath(), currentActivity)
                     .addOnSuccessListener(
                         unused -> {
-                          postUpdateProgress(
-                              updateTask, file.length(), file.length(), UpdateStatus.DOWNLOADED);
+                          postUpdateProgress(file.length(), file.length(), UpdateStatus.INSTALLED);
                           updateTask.setResult();
                         })
                     .addOnFailureListener(
```

Comment:
If we changed this to
.OnSuccessTask(unused -> {
                          postUpdateProgress(file.length(), file.length(), UpdateStatus.INSTALLED);
                          updateTask.setResult();
})
.continueWith( task -> if (task.isComplete() && !task.isSuccessful()) {
        // check the exception status on the task.getException and call postUpdateProgress
      }
      
    Then we might be able to remove the class variables. I don't actually think we need the number of bytes downloaded for installation errors, because if we got to installation we know the download completed so we can just set both totalfilesizebytes and bytes downloaded to  file.size()

Reply:
Update: we chatted  and feel that the cachedTask is fine, but we'd prefer not to store the downloadedbytes as class variables and split logic so that sometimes we use the downloaded bytes from the method parameters while other times we use the downloadedbytes from the class variable